### PR TITLE
Support DistributiveLaw as jbool.Rule

### LINKED ISSUE #1299: BooleanExpressionRule: refactor a logic of the distributive law as a rule for jbool library
Now BooleanExpressionRule applies the `distributive law` manipulating with a String and tokens directly.
The jbool library allows to write a custom rule (`com.bpodgursky.jbool_expressions.rules.Rule`).
Suggestion to refactor a logic for the distributive law as a rule for jbool library

### PR DESCRIPTION
**Pull request by nulls**
### What's done:
- implemented DistributiveLaw extends com.bpodgursky.jbool_expressions.rules.Rule.Rule

This pull request closes #1299

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/BooleanExpressionsRule.kt [hunk 1]
```diff
@@ -350,22 +252,91 @@ class BooleanExpressionsRule(configRules: List<RulesConfig>) : DiktatRule(
...
+            val children = input.children ?: return false
+            if (children.size < 2 || children.any { it !is C }) {
+                return false
+            }
+            return findCommonExpression(children) != null
+        }
+
+        private fun findCommonExpression(children: List<Expression<K>>): Expression<K>? = children.drop(1)
+            .fold(children[0].children) { commons, child ->
+                commons.filter { childResult ->
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/BooleanExpressionsRule.kt:326]: FYI: in Kotlin `filter` creates a new ArrayList, calls `add` and causes performance degradation. Usually it is suggested to use `sequence` for such filtering you have.

Just FYI

**Review by akuleshov7 — APPROVED** [review]
lgtm

File: diktat-rules/src/test/resources/test/paragraph3/boolean_expressions/DistributiveLawExpected.kt [hunk 2]
```diff
@@ -21,4 +21,19 @@ fun some() {
     if (a > 5 && (b > 6 || c > 7 || d > 8)) {
 
     }
+
+    // long case
+    if (a > 5 && ((b > 6 && z > 3) || (c > 7 && y > 4) || (d > 8 && w > 5))) {
```
Comment (akuleshov7) [inline diktat-rules/src/test/resources/test/paragraph3/boolean_expressions/DistributiveLawExpected.kt:26]: First I wanted to suggest not to use braces with `&&` as `a && b || c && d` is the same to `(a && b) || (c && d)`, but now I think that your version is better

**Comment by nulls** [comment]
A tag for `jbool` 1.24 is cut with our change. When a new version will be available, we will get a PR from `renovate bot`

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/BooleanExpressionsRule.kt (hunk 1 above)
Comment (nulls) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/BooleanExpressionsRule.kt:326]: noted, thanks

**Pull request merged**